### DISCUSSION TITLE
Node Embedding using GAE

### DISCUSSION THREAD

**Discussion by elbasir** [pyg-team/pytorch_geometric#5125] [Q&A] [answered]
Hi,

My understanding of GAE is the capability to generate an embedding which represents an entire graph, but assume I have a graph G with N nodes without any labels, I want to extract an embedding for each node in G. Can you guide me on how to get a node embedding using GAE or if there is another unsupervised method which can handle this?

Thanks,

**Comment by rusty1s** [comment]
GAE will produce node embeddings such that nodes that should be connected to each other have a similar embedding while nodes that are far away from each other should also be far away in the embedding space (similar to Node2Vec but based on GNN as encoder). As such, GAE already returns node-level embeddings as part of its encoder.

**Comment by elbasir** [comment]
Thanks for the answer. Assume I have a set of proteins, if I generate a graph of each protein, I pass each graph to GAE and the embedding would be representing the entire protein graph, what about if I have PPI network and I want to generate embedding of each node "protein" here? I checked GAE examples, and what I noticed, it takes the entire graph and generate only one embedding, can you provide me an example how GAE will produce node embeddings in a such scenario as PPI network? I appreciate your assistance.

**Reply by rusty1s** [reply]
I think your understanding is slightly incorrect. As mentioned earlier, GAE will not embed a graph, but will put each node in an embeddding space. You can see this [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/autoencoder.py#L30-L38) where the encoder returns a matrix of shape `[num_nodes, num_embedding_dim]`. As such, GAE already does what you want.

**Comment by elbasir** [comment, marked as answer]
Excuse my incorrect understanding but now it is clear. So, if I want to have an embedding for the entire graph should I aggregate all the `num_embedding_dim` for all nodes in the graph?

**Reply by elbasir** [reply]
In addition to my previous reply, if I have a list of drugs to classify each of them to be either toxic or non-toxic. The `num_features=78` for each drug. Assume I have two drugs to classify them, drug1 has this graph information `Data(x=[31, 78], edge_index=[2, 70], y=[1], target=[1, 546], c_size=[1])` and drug2 has the following graph information `Data(x=[62, 78], edge_index=[2, 136], y=[1], target=[1, 546], c_size=[1])` once they are passed to the GCN model with a `batch_size=2` , I then pass the output of GCN to multiple fully connected layers to make the prediction.

However, this is the output of my GCN `torch.Size([93, 78])` when the two drug graphs are passed in one batch together. How the model could distinguish between them and why it combines the `num_nodes` to be 93 for both drugs? This is really confusing to me. It supposed to return something like `[num_graphs, num_nodes, num_embedding_dim]` to keep each graph output separate from the other graph. Correct me if GCN can distinguish between the two drugs when they are passed together in one batch!

**Reply by rusty1s** [reply]
PyG combines the batch and node dimension into a single dimension to support features of varying sizes without additional memory requirements introduced by padding. As such, 31 nodes + 62 nodes = 93 nodes. Since edges are not shared between the two graphs, there is no information exchange across examples. You can then use the `batch` vector for global readout. Our documentation goes over mini-batching in more detail, so please feel free to take a look.

**Reply by elbasir** [reply]
Thanks for your answer. My final thoughts about getting the graph embedding, should I aggregate the output of GAE layer which is `[num_nodes, num_embedding_dim]`?

**Reply by rusty1s** [reply]
You can aggregate the output to a graph-level representation, but since you never trained against this representation I am not sure how representative of the graph it is.

**Reply by elbasir** [reply]
The reason of getting the embedding representation at the graph level to be used for downstream tasks. Would you recommend something else to generate an embedding of the entire graph? Keep in mind the graph size is not big, max of 5000 nodes per graph.

**Reply by rusty1s** [reply]
What you can also try is Deep Graph Infomax (see `examples/infomax.py`), and compare it to the GAE approach.

**Reply by elbasir** [reply]
Does Deep Graph Infomax retrieve an embedding of an entire graph in this case or it follows the same learning of GAE? I will investigate both of them to have an unsupervised approach for generating a graph representation.

**Reply by rusty1s** [reply]
It actually can both - graph-level representations can be obtained through the summary vector.

**Reply by elbasir** [reply]
BTW, I checked `examples` folder but `infomax.py` does not exist but there are `examples/infomax_inductive.py` and `examples/infomax_transductive.py`, which one should be used?

**Reply by rusty1s** [reply]
Ah right. I suggest to look into the transductive one first.
